File: ovvium-backend/src/main/java/com/ovvium/services/web/controller/bff/v1/transfer/response/bill/BillResponse.java
package com.ovvium.services.web.controller.bff.v1.transfer.response.bill;

import com.ovvium.services.model.bill.Bill;
import com.ovvium.services.web.controller.bff.v1.transfer.response.common.ResourceIdResponse;
import com.ovvium.services.web.controller.bff.v1.transfer.response.customer.CustomerResponse;
import com.ovvium.services.web.controller.bff.v1.transfer.response.customer.EmployeeResponse;
import com.ovvium.services.web.controller.bff.v1.transfer.response.location.LocationResponse;
import com.ovvium.services.web.controller.bff.v1.transfer.response.order.OrderResponse;
import com.ovvium.services.web.controller.bff.v1.transfer.response.user.UserResponse;
import lombok.Getter;

import java.util.List;
import java.util.stream.Collectors;

@Getter
public final class BillResponse extends ResourceIdResponse {

    private final CustomerResponse customer;
    private final List<LocationResponse> locations;
    private final String status;
    private final EmployeeResponse employee;
    private final List<UserResponse> members;
    private final List<OrderResponse> orders;
    private final boolean hasJoinedLocations;

    // FIXME remove and use updateDate instead
    private final long updated;
    private final String creationDate;
    private final String updateDate;

    public BillResponse(Bill bill, CustomerResponse customer, List<UserResponse> members, List<OrderResponse> orders) {
        super(bill);
        this.customer = customer;
        this.locations = bill.getLocations().stream().map(LocationResponse::new).collect(Collectors.toList());
        this.status = bill.getBillStatus().name();
        this.employee = bill.getEmployee().map(EmployeeResponse::new).orElse(null);
        this.members = members;
        this.hasJoinedLocations = bill.hasJoinedLocations();
        this.orders = orders;
        this.updated = bill.getUpdated().toEpochMilli();
        this.creationDate = bill.getCreated().toString();
        this.updateDate = bill.getUpdated().toString();
    }
}

File: ovvium-app/src/model/response/BillResponse.ts
import { LocationResponse } from './LocationResponse';
import { OrderResponse } from './OrderResponse';
import { ResourceIdResponse } from './ResourceIdResponse';
import { UserResponse } from './UserResponse';
import { CustomerResponse } from './CustomerResponse';

export class BillResponse extends ResourceIdResponse {
    customer: CustomerResponse;
    locations: LocationResponse[];
    status: string;
    members: UserResponse[];
    orders: OrderResponse[];
    hasJoinedLocations: boolean;
    updated: number;

	constructor(billResponse: BillResponse = {} as BillResponse) {
        super(billResponse.id);
        this.customer = billResponse.customer;
        this.locations = billResponse.locations;
        this.status = billResponse.status;
        this.members = billResponse.members;
        this.orders = billResponse.orders;
        this.hasJoinedLocations = billResponse.hasJoinedLocations;
        this.updated = billResponse.updated;
    }
}